### PR TITLE
Allow specially named uninterpreted functions to introduce SMT-LIB di…

### PR DESCRIPTION
**Pull request by martin-cs**
…rectly.

This is a potential code injection to the SMT solver but that shouldn't be a massive issue.  It's a bit of a hack but allows fast prototyping of solver features.

### CODE REVIEW

File: src/solvers/smt2/smt2_conv.cpp [hunk 1]
```diff
@@ -8,6 +8,8 @@ Author: Daniel Kroening, kroening@kroening.com
 
 #include <cassert>
 
+#include <string.h>
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_conv.cpp]: Use cstring instead of string.h

File: src/solvers/smt2/smt2_conv.cpp [hunk 2]
```diff
@@ -950,6 +952,82 @@ std::string smt2_convt::convert_identifier(const irep_idt &identifier)
...
+    size_t length = strlen(escapeName);
+
+    std::string id(id2string(to_symbol_expr(symbol).get_identifier()));
+
+    if (id.length() >= length)
+      return id.compare(0, length, escapeName, length) == 0;
+  }
+
+  return false;
+}
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_conv.cpp]: Use has_prefix(...) ?

File: src/solvers/smt2/smt2_conv.cpp [hunk 3]
```diff
@@ -950,6 +952,82 @@ std::string smt2_convt::convert_identifier(const irep_idt &identifier)
...
+
+std::string smt2_convt::convert_function_name(const function_application_exprt &fa) {
+  symbol_exprt function(to_symbol_expr(fa.function()));
+
+  if (uf_is_escaped_smt(function))
+  {
+    exprt firstArgument(fa.arguments()[0]);
+
+    // We want a pointer to a string_constant
+    address_of_exprt addr(to_address_of_expr(firstArgument));
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_conv.cpp]: Use const references instead of constructing new objects.

File: src/solvers/smt2/smt2_conv.cpp [hunk 4]
```diff
@@ -4586,10 +4681,31 @@ void smt2_convt::find_symbols(const exprt &expr)
...
+	  << "| ";
+
+      if(expr.type().id()==ID_code)
+      {
+	code_typet functionType(to_code_type(expr.type()));
+	out << "( ";
+	code_typet::parameterst parameters(functionType.parameters());
+	for (code_typet::parameterst::const_iterator i = parameters.begin();
+	     i != parameters.end();
+	     ++i)
```
Comment (tautschnig) [inline src/solvers/smt2/smt2_conv.cpp]: ranged for

**Comment by thomasspriggs** [comment]
I am going to close this PR as it appears that work on it has been inactive for an extended period of time. This is not any judgement on how worthwhile this work is. This is part of an effort to reduce the number of open PRs which are not being actively worked on. If you still think it would be worthwhile to get this merged then the branch should be re-based on the latest version of develop and the PR re-opened.

**Pull request closed without merging**